### ISSUE TITLE
Revisions to Declaration.cs and related classes/enums like accessibility

### ISSUE THREAD

**Issue by ThunderFrame** [rubberduck-vba/Rubberduck#4100]
There are a number of issues with Delaration.cs. This issue serves as a placeholder to discuss and revise this core class and its associated classes and enums.

Declarations are central to many features of RD, namely the resolver and inspections, so changes to these files will entail many changes in other parts of RD. Let's think very carefully about how we revise these critical files.

### Accessibility.cs
The current enum has 6 members, despite VBA only having 3 levels of accessibility
```c#
    public enum Accessibility
    {
        Private = Source.Accessibility.Private,
        Friend = Source.Accessibility.Friend,
        Global = Source.Accessibility.Global,
        Implicit = Source.Accessibility.Implicit,
        Public = Source.Accessibility.Public,
        Static = Source.Accessibility.Static
}
```
VBA has only 3 levels of accessibility: `Public`, `Friend` and `Private`.
- `Global` is an alias (and legacy) term for `Public`, and should perhaps be a `AccessibilityDeclaredUsingLegacyGlobalKeyword` property of the Declaration?
- `Static` is not related to accessibility at all, and should be a property of the declaration.
- `Implicit` should be an `AccessibilityIsExplicit` property of the declaration, not a type of accessibility.

### Declaration.cs
So, `Declaration` needs some new properties:
**Accessibility**
As per above discusssion:
- `bool AccessibilityDeclaredUsingLegacyGlobalKeyword` indicates a declaration with `Public` accessibility was declared using the legacy keyword `Global`
- `bool AccessibilityIsExplicit` indicates the accessibility was explicitly declared.

**Static**
Remembering that `Static` in VBA relates to the lifetime of variables. Procedures declared as static have all of their constituent variables (regardless of whether each variable is declared static) retain their values between calls. Variables that are explicitly declared static retain their values, even if the procedure is not declared static.
- `bool DeclarationIsStatic` indicates whether the declaration is static
- `bool DeclarationIsExplicitlyStatic` indicates whether the declaration is explicitly declared static. For example, a static procedure would be explicitly declared, whereas the constituent variables of a static procedure might be implicitly static.

**Comment by MDoerner** [comment]
I agree except for two points:
1. I do not think we need a property to state whether that global has been used to declare the accessibility. That is only interesting for an inspection, which can simply look at the `context` or be implemented as parse tree inspection.
2. I think we should have an `IsStatic` property instead of `DeclarationIsExplicitlyStatic`.

**Comment by BZngr** [comment]
In general, the RD codebase has a minority of classes defined without an associated interface.  The design, implementation, and use of `Declaration` is part of that minority.  As a result, `Declaration` and the codebase loses out on the opportunities afforded objects that conform to the Liskov Substitution Principle.  There's a lot I do not know about RD's history and discussions (formal or otherwise) surrounding the `Declaration` class.  So, if this is a previously resolved discussion/question, I apologize.  If not, I would be interested getting reactions to the suggestion of creating an `IDeclaration` interface. 

**Justification:**
There are already specific use cases in the RD codebase that could benefit from an `IDeclaration` interface.  I'm most familiar with refactorings, so I'll stick with examples from there for now.  Fundamentally, they all boil down to supporting the possibility of taking a declaration's read-only attributes and making them writable using the Decorator pattern other mechanisms. 

_ICodeBuilder/CodeBuilder:_ 
For the most part, `CodeBuilder` generates blocks of VBA code that are introduced as new `Declaration(s)`.  Consequently, much of the `CodeBuilder` interface requires a 'prototype' declaration and a set of attributes to use when generating the block of code.  With an `IDeclaration` interface, the `CodeBuilder` could operate directly on an `IDeclaration` reference (basically a _CodeBuilderModel_)  that represents the _new_ `Declaration`.  

Refactorings that use directly or indirectly`ICodeBuilder` to introduce new declarations:  

- `ExtractInterfaceRefactoring`
- `ImplementInterfaceRefactoring`
- `AddInterfaceImplementationsRefactoring`
- `EncapsulateFieldRefactoring` 

_ConflictDetection:_
The greatest benefit (at least for refactorings) occurs with regard to conflict detection IMO.  For the above refactorings and `RenameRefactoring`, a fair amount of code and logic is dedicated to allowing the refactoring's conflict detector to make evaluations as though the codebase already contains all new/proposed or renamed declarations.  The ability to create something like a `NewDeclarationProxy<IDeclaration>` would be both intuitive and useful.

_Testing:_
`Mock<IDeclaration>`...TBH, I don't think I've run across a testing scenario where I _really_ _really_ wished I could do this...but it would become possible.  It is likely that scenarios where mocking a `Declaration` is useful - will become self-evident.  For example, the `CodeBuilder` tests could probably all be done using `Mock<IDeclaration>` and never require parsing a single line of VBA code.

**Comment by Vogel612** [comment]
Historically declarations are read-only because we don't have a rewriting mechanism that can deal with declarations. 
Additionally the Nature of declarations allows for certain guarantees that may be lost when an interface exposes writability.

That being said the current state of declarations is not something that we want to have, if only because `Declaration` deals with so many concerns at once

**Comment by BZngr** [comment]
My intent is that `IDeclaration` would match the default interface of `Declaration` exactly.  All read-only properties remain read-only.  The notion of writable properties are for imagined scenarios like `IDeclarationProxy : IDeclaration` where the proxy interface extends and overrides `IDeclaration` to provide read/write mutators and functions for the purposes of creating a proxy for specific scenarios like a new or renamed `Declaration`.

**Comment by BZngr** [comment]
[DefaultMemberUsage.xlsx](https://github.com/rubberduck-vba/Rubberduck/files/5484505/DefaultMemberUsage.xlsx)

Collated some usage data regarding the default `Declaration` interface.  Perhaps interesting...perhaps not.
